pkg/apis/harvester.cattle.io/v1alpha1/backup.go

```diff
@@ -0,0 +1,251 @@
+	VirtualMachineRestoreDelete DeletionPolicy = "delete"
+
+	// VirtualMachineRestoreRetain causes the VirtualMachineRestore deleted resources like dataVolume or PVC to be retained
+	VirtualMachineRestoreRetain DeletionPolicy = "retain"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:shortName=vmbackup;vmbackups,scope=Namespaced
+// +kubebuilder:printcolumn:name="SOURCE_KIND",type=string,JSONPath=`.spec.source.kind`
+// +kubebuilder:printcolumn:name="SOURCE_NAME",type=string,JSONPath=`.spec.source.name`
+// +kubebuilder:printcolumn:name="READY_TO_USE",type=boolean,JSONPath=`.status.readyToUse`
+// +kubebuilder:printcolumn:name="AGE",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:printcolumn:name="ERROR",type=date,JSONPath=`.status.error.message`
+
+type VirtualMachineBackup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualMachineBackupSpec `json:"spec"`
+
+	// +optional
+	Status *VirtualMachineBackupStatus `json:"status,omitempty"`
+}
+
+type VirtualMachineBackupSpec struct {
+	Source corev1.TypedLocalObjectReference `json:"source"`
+}
+
+// VirtualMachineBackupStatus is the status for a VirtualMachineBackup resource
+type VirtualMachineBackupStatus struct {
+	// +optional
+	SourceUID *types.UID `json:"sourceUID,omitempty"`
+
+	// +optional
+	VirtualMachineBackupContentName *string `json:"virtualMachineBackupContentName,omitempty"`
+
+	// +optional
+	ReadyToUse *bool `json:"readyToUse,omitempty"`
+
+	// +optional
+	Error *Error `json:"error,omitempty"`
+
+	// +optional
+	Conditions []Condition `json:"conditions,omitempty"`
+}
+
+// Error is the last error encountered during the snapshot/restore
+type Error struct {
+	// +optional
+	Time *metav1.Time `json:"time,omitempty"`
+
+	// +optional
+	Message *string `json:"message,omitempty"`
+}
+
+// VirtualMachineBackupContent contains the snapshot data
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:shortName=vmbackupcontent;vmbackupcontents,scope=Namespaced
+// +kubebuilder:printcolumn:name="READY_TO_USE",type=boolean,JSONPath=`.status.readyToUse`
+// +kubebuilder:printcolumn:name="AGE",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:printcolumn:name="ERROR",type=string,JSONPath=`.status.error.message`
+
+type VirtualMachineBackupContent struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualMachineBackupContentSpec `json:"spec"`
+
+	// +optional
+	Status *VirtualMachineBackupContentStatus `json:"status,omitempty"`
+}
+
+// VirtualMachineBackupContentSpec is the spec for a VirtualMachineBackupContent resource
+type VirtualMachineBackupContentSpec struct {
+	VirtualMachineBackupName *string `json:"virtualMachineBackupName,omitempty"`
+
+	Source SourceSpec `json:"source"`
```

Comment:
Can we use VirtualMachineSpec here

Reply:
the `sourceSpec` has contained the VirtualMachineSpec